codegen/fragment.go

```diff
@@ -1,95 +1,277 @@
+	Module *ir.Module
+}
 
-	fragment.Lines = make([]IrLiner, 0)
-	fragment.identifierTracker = identifierTracker
-	fragment.ResultType = resultType
-	fragment.result = ""
+func NewModuleFragment(module *ir.Module) *ModuleFragment {
+	if module == nil {
+		module = ir.NewModule()
+	}
 
-	return &fragment
+	return &ModuleFragment{Module: module}
 }
 
-// Append another code fragment
-func (fragment *Fragment) Append(frag *Fragment) {
-	fragment.Lines = append(fragment.Lines, frag.Lines...)
+func (moduleFragment *ModuleFragment) Append(fragment Fragment) {
+	switch f := fragment.(type) {
+	case *ModuleFragment:
+		panic("Cannot append module to module")
+	case *FunctionsFragment:
+		moduleFragment.Module.Funcs = append(moduleFragment.Module.Funcs, f.Functions...)
+
+		for _, function := range f.Functions {
+			function.Parent = moduleFragment.Module
+		}
+	case *BlocksFragment:
+		panic("Cannot append blocks to module")
+	case *InstructionsFragment:
+		panic("Cannot append instructions to module")
+	}
+}
+
+func (moduleFragment *ModuleFragment) AppendWithResultPropogation(fragment Fragment) {
+	moduleFragment.Append(fragment)
 }
 
-func (fragment *Fragment) AppendBefore(frag *Fragment) {
-	fragment.Lines = append(frag.Lines, fragment.Lines...)
+func (moduleFragment *ModuleFragment) GetResultType() ResultType {
+	return VOID
 }
 
-func (fragment *Fragment) AppendLines(lines ...IrLiner) {
-	fragment.Lines = append(fragment.Lines, lines...)
+func (moduleFragment *ModuleFragment) GetResult() value.Value {
+	return nil
 }
 
-// AddInstruction adds an instruction
-func (fragment *Fragment) AddInstruction(format string, arguments ...interface{}) {
-	operation := fmt.Sprintf(format, arguments...)
+type FunctionsFragment struct {
+	Functions       []*ir.Func
+	currentFunction *ir.Func
+}
+
+func NewFunctionsFragment() *FunctionsFragment {
+	functions := make([]*ir.Func, 0)
+
+	return &FunctionsFragment{Functions: functions, currentFunction: nil}
+}
 
-	instruction := &Instruction{Result: "", Operation: operation}
+func (functionsFragment *FunctionsFragment) AddFunc(function *ir.Func) {
+	if function == nil {
+		panic("Cannot add nil function")
+	}
 
-	fragment.Lines = append(fragment.Lines, instruction)
+	functionsFragment.Functions = append(functionsFragment.Functions, function)
+	functionsFragment.currentFunction = function
 }
 
-// AddOperation adds an operation, and return the result
-func (fragment *Fragment) AddOperation(format string, arguments ...interface{}) string {
-	operation := fmt.Sprintf(format, arguments...)
+func (functionsFragment *FunctionsFragment) Append(fragment Fragment) {
+	switch f := fragment.(type) {
+	case *ModuleFragment:
+		panic("Cannot append module to function")
+	case *FunctionsFragment:
+		functionsFragment.Functions = append(functionsFragment.Functions, f.Functions...)
+
+		functionsFragment.currentFunction = f.currentFunction
+	case *BlocksFragment:
+		blocks := functionsFragment.currentFunction.Blocks
```

Comment:
A `FunctionsFragment` should not have an empty array of functions at this point. Maybe do a `panic`